src/main/java/banking/Bank.java

```diff
@@ -0,0 +1,68 @@
+package banking;
+
+import banking.database.AccountDao;
+import banking.model.Account;
+import banking.model.Card;
+import banking.util.CardGenerator;
+import banking.util.CardValidator;
+
+import java.util.Optional;
+
+public class Bank {
+
+    private final AccountDao accountDao;
+    private final CardValidator cardValidator;
+    private final CardGenerator cardGenerator;
+
+    public Bank(AccountDao accountDao, String BINumber) {
+        this.accountDao = accountDao;
+        this.cardValidator = new CardValidator();
+        this.cardGenerator = new CardGenerator(cardValidator, BINumber);
+    }
+
+    public Card registerAccount() {
+        Card card;
+
+        do {
+            card = cardGenerator.generate();
+        } while (exists(card.cardNumber()));
+
+        accountDao.save(card);
+        return card;
+    }
+
+    public Optional<Account> signIn(String cardNumber, String cardPIN) {
+        return accountDao.get(cardNumber, cardPIN);
+    }
+
+    public boolean exists(String cardNumber) {
+        return accountDao.get(cardNumber).isPresent();
+    }
+
+    public void addIncome(String cardNumber, int income) {
+        accountDao.update(cardNumber, income);
+    }
+
+    public void doTransfer(String from, String to, int amount) {
+        addIncome(from, -amount);
+        addIncome(to, amount);
+    }
+
+    public void removeAccount(int id) {
+        accountDao.delete(id);
+    }
+
+    public void checkCardNumber(String cardNumber) throws Exception {
```

Comment:
Maybe it’s better to move this functionality into cardValidator? or it’s for other purposes?

Reply:
CardValidator API was designed to only do two things: generate last digit for the card number and check if the last digit is correct for passed number. And in this method `checkCardNumber` we are checking if that card is valid and if our bank already contains that card. So I'd leave it as it is, but if you insist I will move it into cardValidator